bootloader/src/old_machine_bootloader_windows.c

```diff
@@ -0,0 +1,330 @@
+#ifdef _WIN32
+
+#include <winsock2.h>
+#include <windows.h>
+#include <wininet.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <iphlpapi.h>
+#include <VersionHelpers.h>
+#include <wchar.h>
+
+#pragma comment ( lib, "wininet" )
+#pragma comment ( lib, "Wininet.lib" )
+
+#define XP_OR_LOWER "xp_or_lower"
+#define VISTA "vista"
+#define VISTASP1 "vista_sp1"
+#define VISTASP2 "vista_sp2"
+#define WINDOWS7 "windows7"
+#define WINDOWS7SP1 "windows7_sp1"
+#define WINDOWS8_OR_GREATER "windows8_or_greater"
+
+#define BOOTLOADER_SERVER_PORT 5001
+#define ISLAND_SERVER_PORT L":5000"
+
+void error(const char *msg) { perror(msg); exit(1); }
+
+struct requestData {
+    char* IPstring;
+    char* hostname;
+    char* osVersion;
+    BOOL tunnelUsed;
+    char* tunnel;
+};
+
+wchar_t* getRequestDataJson(struct requestData reqData){
+    char tunnel[26];
+    if(reqData.tunnelUsed){
+        snprintf(tunnel, 26, "%s%s%s", "\"", reqData.tunnel, "\"");
+    } else {
+        strcpy_s(tunnel, _countof(tunnel), "false");
+    }
+    // Concatenate into string for post data
+    char* buf = malloc(sizeof(char) * (2000));
+    snprintf(buf,
+             (sizeof(char) * (2000)),
+             "{\"system\":\"%s\", \"os_version\":\"%s\", \"hostname\":\"%s\", \"tunnel\":%s, \"ips\": [\"%s\"]}",
+             "windows",
+             reqData.osVersion,
+             reqData.hostname,
+             tunnel,
+             reqData.IPstring);
+    wchar_t* requestDataJson = (wchar_t*)malloc(sizeof(wchar_t) * (strlen(buf)+1));
+    mbstowcs(requestDataJson, buf, strlen(buf)+1);
+    wprintf(L"Request data: %ls\n", requestDataJson);
+    return requestDataJson;
+}
+
+char** getIpAddresses(int maxSize, int *addrCount, char** hostname){
+    char** IPs = malloc(maxSize * sizeof(char*));
+    int j = 0;
+
+    DWORD Err;
+    PFIXED_INFO pFixedInfo;
+    DWORD FixedInfoSize = 0;
+
+    PIP_ADAPTER_INFO pAdapterInfo, pAdapt;
+    DWORD AdapterInfoSize;
+    PIP_ADDR_STRING pAddrStr;
+
+    //
+    // Get the main IP configuration information for this machine using a FIXED_INFO structure
+    //
+    if ((Err = GetNetworkParams(NULL, &FixedInfoSize)) != 0)
+    {
+        if (Err != ERROR_BUFFER_OVERFLOW)
+            error("GetNetworkParams sizing failed\n");
+    }
+
+    // Allocate memory from sizing information
+    if ((pFixedInfo = (PFIXED_INFO) GlobalAlloc(GPTR, FixedInfoSize)) == NULL)
+        error("Memory allocation error\n");
+
+    if ((Err = GetNetworkParams(pFixedInfo, &FixedInfoSize)) == 0)
+    {
+        *hostname = (char *)malloc(strlen(pFixedInfo->HostName)+1);
```

Comment:
Check return value. Always

Reply:
return of which function? If `GetNetworkParams` fails runtime will skip this block

Reply:
If error won't quit (bad practice) then rest of blocks will execute.